docs(refactor): updates the cruise control rebalance concepts

PaulRMellor opened the pull request:
**Documentation**

Refactor and refresh of Cruise Control concepts
- Adds new optimization proposal process flow description and diagram, including description of partition reassignment commands
- Less verbose and more direct intro and concepts
- Removes three overview files to create a single "components and features" file for goals and proposals concepts
- Consolidates related conceptual information (goals, proposals) into single sections
- Retitled sections to provide more direction to readers from ToC

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [x] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md
- [ ] Supply screenshots for visual changes, such as Grafana dashboards

fvaleri commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -2,54 +2,501 @@
+
+* *Hard goals* are preset and mandatory for a proposal to succeed.
+* *Soft goals* are optional and can be set aside if hard goals are met.
+* *Main goals* are inherited from Cruise Control, some preset as hard goals, used by default.
+* *Default goals* are the same as main goals by default but customizable.
```
As the name suggests, the Analyzer's `default.goals` is what it's used by default when requesting a proposal, and it is equal to the Analyzer's `goals` if the user does not set `default.goals`. 

The `goals` property is customizable in case the user wants to restrict the supported goals to a subset of what CruiseControl provides by default. This means that, if a goal is not present in Analyzer's `goals`, the it cannot be used in Analyzer's `default.goals` and proposal's `goals`.

The rest of the documentation seems to match my understanding. Do you think we can improve the wording a bit here?

PaulRMellor replied:
Thanks. I've tried to make this clearer and keep succinct.

fvaleri commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -2,54 +2,501 @@
+* *Soft goals* are optional and can be set aside if hard goals are met.
+* *Main goals* are inherited from Cruise Control, some preset as hard goals, used by default.
+* *Default goals* are the same as main goals by default but customizable.
+* *User-provided goals* are a subset of default goals configured for specific proposals.
```
"User provided" (here and in other places) is also a bit confusing for me, because all goals configurations can be user provided. What about `KafkaRebalance` goals?

PaulRMellor replied:
Thanks @fvaleri  
What about "proposal-specific"?  I think bringing a resource name into the list of types might be just as confusing.
@kyguy -- do you have any thoughts on this?

kyguy replied:
`proposal-specific` sounds good to me too

fvaleri commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -2,54 +2,501 @@
+. Load Monitor collects the metrics from Kafka brokers, including CPU, disk, and network utilization data.
+. Anomaly Detector continuously monitors the collected metrics to identify anomalies, such as broker failures or disk capacity issues, that could impact cluster stability.
+. Analyzer processes the collected metrics and constructs a _workload model_ of the current state of the Kafka cluster. 
+It generates an optimization proposal based on configured goals (and constraints) such as balancing partition distribution across brokers, which is sent to the status of the `KafkaRebalance` resource.
```
```suggestion
Based on configured goals and capacities, it generates an optimization proposal for balancing partition distribution across brokers, which is finally reflected in the status of the `KafkaRebalance` resource.
```

fvaleri commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -2,54 +2,501 @@
+Intra-broker rebalancing moves data between disks on the same broker when you are using a JBOD storage configuration.
+Such information can be useful even if you don't go ahead and approve the proposal.
+
+You might reject an optimization proposal, or delay its approval, because of the additional load on a Kafka cluster when rebalancing.
```
We might want to add that if the proposal is too old, the cluster load may have changed significantly, so it would be better to request a new proposal.

fvaleri commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -2,54 +2,501 @@
+----
+
+The proposal will also move 24 partition leaders to different brokers.
+This requires a change to the ZooKeeper configuration, which has a low impact on performance.
```
```suggestion
This requires a change to the cluster metadata, which has a low impact on performance.
```

fvaleri reviewed (commented):
Hi @PaulRMellor, thanks. I must say that our CC doc is quite good :)

I left some comments for your consideration.

kyguy commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -2,54 +2,501 @@
+
+They are categorized as follows:
+
+* *Hard goals* are preset and mandatory for a proposal to succeed.
```
What do we mean by "mandatory" here? That the hard goals must be set? Or "must be satisfied"? The latter is a key distinction of hard goals.

PaulRMellor replied:
change to "must be satisfied"

kyguy commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -2,54 +2,501 @@
+They are categorized as follows:
+
+* *Hard goals* are preset and mandatory for a proposal to succeed.
+* *Soft goals* are optional and can be set aside if hard goals are met.
```
Even in the original doc, the phrase "They can be set aside if it means that all hard goals are met" sounds as if they soft goals are ignored or unimportant if hard goals are met. However, even if the hard goals are met, the soft goals are still important. Although the soft goals are best effort and will not prevent a optimization proposal from being generated, they are still taken it account when create a optimization proposal.

I would keep the original bullets concerning "hard goals" and "soft goals" but fix the line "They can be set aside if it means that all hard goals are met." to show that soft goals are "best effort"

PaulRMellor replied:
Thanks. Hopefully this is clearer to the intention:

* *Soft goals* are objectives that are prioritized during optimization as much as possible, without preventing a proposal from being created if all hard goals are satisfied.

kyguy commented on documentation/assemblies/cruise-control/assembly-cruise-control-concepts.adoc:
```diff
@@ -5,17 +5,21 @@
 
-include::../../modules/cruise-control/con-cruise-control-description.adoc[leveloffset=+1]
+[role="_abstract"]
+Cruise Control is an open-source system for Kafka that monitors broker loads and rebalances partitions to optimize use of resources across the cluster.
```
I see we use the word  "system" in the original doc before these changes but is it the correct word? Wouldn't "application" be a better fit here?
```suggestion
Cruise Control is an open-source application designed to run alongside Kafka to help optimize use of cluster resources by:
* Monitoring cluster workload
* Rebalancing partitions based on predefined constraints
```

kyguy commented on documentation/assemblies/cruise-control/assembly-cruise-control-concepts.adoc:
```diff
@@ -5,17 +5,21 @@
-include::../../modules/cruise-control/con-cruise-control-description.adoc[leveloffset=+1]
+[role="_abstract"]
+Cruise Control is an open-source system for Kafka that monitors broker loads and rebalances partitions to optimize use of resources across the cluster.
+Rebalances help with running a more balanced Kafka cluster that uses brokers more efficiently.
```
Since we use Cruise Control for more than just rebalances, for example topic replication factor changes, I would suggest we keep the original line here

```suggestion
The operations help with running a more balanced Kafka cluster that uses broker pods more efficiently.
```

kyguy commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -2,54 +2,501 @@
+The classification of hard and soft goals is fixed in Cruise Control code and cannot be changed.
+
+A proposal meeting all hard goals is valid, even if it violates some soft goals. 
+Cruise Control prioritizes satisfying hard goals and then maximizes soft goals.
```
```suggestion
Cruise Control prioritizes satisfying hard goals and then prioritizes satisfy soft goals in the order they are listed.
```

kyguy commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -2,54 +2,501 @@
+
+=== Hard and soft goals
+
+Hard goals are mandatory and must be satisfied for optimization proposals to be valid. 
```
```suggestion
Hard goals are mandatory and must be satisfied for optimization proposals to be generated. 
```

kyguy commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -2,54 +2,501 @@
+* To specify hard goals, list them in `hard.goals`.
+* To exclude a hard goal, ensure it's not in either `default.goals` or `hard.goals`.
+
+Increasing the number of configured hard goals will reduce the likelihood of Cruise Control generating valid optimization proposals.
```
It'll reduce the likelihood of generating an optimization proposal period. From what I understand there is no concept of an invalid optimization proposal. If the hard goals cannot be satisfied, a proposal will not be generated
```suggestion
Increasing the number of configured hard goals will reduce the likelihood of Cruise Control generating optimization proposals.
```

PaulRMellor commented:
Thanks for the reviews @fvaleri  and @kyguy 
I've addressed all the comments except the suggestion to change  "user-defined goals" to something else.
I agree it's not clear, but I'm not sure changing to "KafkaRebalance goals" is the way:
https://github.com/strimzi/strimzi-kafka-operator/pull/10810#discussion_r1832874907

fvaleri commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -28,16 +29,18 @@ They are categorized as follows:
 * *Soft goals* are optional and can be set aside if hard goals are met.
-* *Main goals* are inherited from Cruise Control, some preset as hard goals, used by default.
-* *Default goals* are the same as main goals by default but customizable.
+* *Main goals* are inherited from Cruise Control and typically used as the default set of goals, with some preset as hard goals.
+The goals can be customized to specify a subset of all available goals provided by Cruise Control.
```
What about:

```
* *Main goals* contains the list of goals that can be used by the Cruise Control operations (all built-in goals by default).
This list can be customized to specify a subset of all the available goals, which may include custom goals.
```

My intention is to avoid giving the impression that you should update this configuration to change default goals. You can, but it would be an indirect update, and your configuration may be lost if someones arrives and sets `default.goals`.

fvaleri reviewed (approved):
LGTM. Thanks.

ppatierno commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -2,54 +2,505 @@
+You specify the new brokers as a list using the `spec.brokers` property of the `KafkaRebalance` custom resource.
+
+`remove-brokers` mode:: The `remove-brokers` mode is used before scaling down a Kafka cluster by removing one or more brokers.
+If you scale down a Kafka cluster, brokers are shut down even if they host replicas.
```
this is actually not the case by default. We have now I check which prevents you to scale down if the brokers are hosting partitions, so that you need to run rebalancing first. There is a `strimzi.io/skip-broker-scaledown-check` annotation which can be set to true to skip this check (but it's disabled by default as just explained).

PaulRMellor replied:
Right. I've updated this and added a note that references the section on skipping checks on scale-down operations

ppatierno commented on documentation/shared/images/kafka-concepts-cruise-control.png:
Anomaly Detector and Analyzer are not called one after the other, the picture is misleading. They work in parallel by using the load monitor.
Also, to get the status there is no direct communication with the Analyzer but there is an additional REST API component which is used for that by our operator.
I would re-arrange the picture by referring to the official one in the Cruise Control repository here: https://github.com/linkedin/cruise-control/wiki/Overview
Of course we don't need all the details but at least fixing the above comments.

PaulRMellor commented:
Thanks for the latest comments @kyguy and @ppatierno 
I've updated the diagram and addressed all comments apart from this https://github.com/strimzi/strimzi-kafka-operator/pull/10810#discussion_r1836090597
Looking for guidance

kyguy commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -2,54 +2,505 @@
+Some main goals are preset as hard goals.
+
+To simplify configuration, use the inherited main goals unless you need to exclude specific goals from `KafkaRebalance` resources. 
+You can adjust the priority order in the default optimization goals configuration.
```
To my understanding yes, the goals are listed in order of priority, so the first goal listed in the configuration will have higher priority than the second goal list

ppatierno replied:
Ok. So I think we should clarify here (and everywhere goals priority is mentioned) that the priority is based on the order. Not sure if "priority order" is used for this purpose but it's not clear to me.

fvaleri replied:
We have to admit that CC naming doesn't help much here.

IMO, this is the default inter-broker goals priority list:

https://github.com/linkedin/cruise-control/blob/main/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/constants/AnalyzerConfig.java#L312-L327

To change it, the user can configure Kafka `default.goals` or KafkaRebalance `goals` with a different order. Only goals listed in Kafka `goals` can be used in the previous configuration (DEFAULT_DEFAULT_GOALS is a subset of DEFAULT_GOALS).

- Do you agree?
- Are we exposing Kafka `goals` because we intend to eventually support custom goals?

kyguy commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -2,54 +2,508 @@
+
+They are categorized as follows:
+
+* *Main goals* are a list of goals supported by the Cruise Control instance and used in its operations. 
```
```suggestion
* *Main goals* are a list of goals supported by the Cruise Control instance that can be used in its operations. 
```

ppatierno replied:
I agree, not sure that "main" makes much sense

PaulRMellor replied:
Yeah. Supported is much better. Changed.

kyguy commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -2,54 +2,508 @@
+They are categorized as follows:
+
+* *Main goals* are a list of goals supported by the Cruise Control instance and used in its operations. 
+By default, this list includes all built-in goals. 
```
What do we mean by "built-in" here? All the goals that come included/shipped with upstream Cruise Control?

kyguy commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -2,54 +2,508 @@
+For a goal to be used in other categories, such as default or hard goals, it must first be listed in main goals. 
+To prevent a goal’s usage, remove it from this list.
+* *Hard goals* are preset and must be satisfied for a proposal to succeed.
+* *Soft goals* are objectives that are prioritized during optimization as much as possible, without preventing a proposal from being created if all hard goals are satisfied.
```
```suggestion
* *Soft goals* are objectives that are prioritized during optimization as much as possible, but won't prevent the generation of a proposal if not satisfied. 
```

kyguy replied:
IIRC, at this time, both hard goals and soft goals are preset by upstream Cruise Control. Basically, whatever is not a `hard.goal` (whatever `goals` not listed in the default `hard.goals` config) are soft goals. 

I am not completely sure why Cruise Control offers this `hard.goals` list in the first place if the goal type is hard coded/preset by upstream Cruise Control. Listing a goal in the `hard.goals` list does not change it into a hard goal from what I remember (but my memory is hazy so let's leave this information out for now, I just wanted people to be aware of it)

I bring this up because it may be worth mentioning that soft goals are preset as well

kyguy commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -2,54 +2,508 @@
+
+=== Hard and soft goals
+
+Hard goals are mandatory and must be satisfied for optimization proposals to be generated. 
```
What do we mean by "mandatory" here?

PaulRMellor replied:
I've removed that.

kyguy commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -2,54 +2,508 @@
+
+=== Main goals
+
+Main goals are predefined and available to all users in Cruise Control. 
```
```suggestion
Main goals are predefined and available to use for generating Cruise Control optimization proposals. 
```

kyguy commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -2,54 +2,508 @@
+A proposal meeting all hard goals is valid, even if it violates some soft goals. 
+
+For example, a soft goal might be to evenly distribute a topic's replicas. 
+Cruise Control will ignore this if it conflicts with hard goals.
```
I wouldn't say the Cruise Control will ignore the soft goal, it just continue to generate an optimization proposal even if the soft goal isn't completely satisfied

kyguy commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -2,54 +2,508 @@
+Cruise Control uses default goals to generate an optimization proposal. 
+You can override default goals by setting proposal-specific optimization goals in a `KafkaRebalance` resource.
+
+If `default.goals` is not specified in the Cruise Control deployment configuration, main goals are used as default goals. 
```
```suggestion
If `default.goals` is not specified in the Cruise Control deployment configuration, Strimzi will configure the `default.goals` to the list of goals specified in `main.goals`. 
```

PaulRMellor replied:
I think we mean `config.goals` not `main.goals`?

kyguy commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -2,54 +2,508 @@
+You can override default goals by setting proposal-specific optimization goals in a `KafkaRebalance` resource.
+
+If `default.goals` is not specified in the Cruise Control deployment configuration, main goals are used as default goals. 
+The optimization proposal is then generated using these main goals.
```
```suggestion
The optimization proposal based on this goals list will then be generated and cached.
```

kyguy commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -2,54 +2,508 @@
+
+=== Proposal-specific goals
+
+Proposal-specific optimization goals narrow down the configured default goals for specific optimization proposals.
```
Not necessarily "narrows down", the configured default goals list could be smaller than proposal-sepcific goals list.
```suggestion
Proposal-specific optimization goals allow creating optimization proposals based on a specific list of goals.
```

ppatierno commented on documentation/shared/images/kafka-concepts-cruise-control.png:
It's better but there is not really a way to approve/reject a proposal, so from our perspective it's more a "run rebalancing", "execute rebalancing", "apply proposal" .... I will leave on you the best naming. It's more like the "Rebalance" you have between the executor and the cluster.

PaulRMellor replied:
Maybe just having the flow as "rebalance" that way makes it simpler to follow.

PaulRMellor commented:
Thanks for the latest @kyguy and @ppatierno 
I've changed "main" to "supported" goals. (Much better)
Diagram now shows "Rebalance" as flow towards CC

ppatierno reviewed (approved):
LGTM.

kyguy commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -2,54 +2,508 @@
+Some supported goals are preset as hard goals.
+
+To simplify configuration, use the inherited supported goals unless you need to exclude specific goals from `KafkaRebalance` resources. 
+You can adjust the priority order in the default optimization goals configuration.
```
Since this is about the `default.goals` list, maybe we should move this line to the "Default goals" section

kyguy commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -2,54 +2,508 @@
+Goals not listed as supported goals cannot be used in Cruise Control operations.
+Some supported goals are preset as hard goals.
+
+To simplify configuration, use the inherited supported goals unless you need to exclude specific goals from `KafkaRebalance` resources. 
```
This  seems to imply that users should edit the `goals` list if they want to exclude goals from the `KafkaRebalance` resource. I would remove this line

kyguy commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -2,54 +2,508 @@
+Goals not listed as supported goals cannot be used in Cruise Control operations.
+Some supported goals are preset as hard goals.
+
+To simplify configuration, use the inherited supported goals unless you need to exclude specific goals from `KafkaRebalance` resources. 
```
This line seems to imply that if we don't want specific inherited supported goals from being used by `KafkaRebalance` resources we can update the supported goal list. However, this could potentially cause problems with the the `default.goals` list and other lists. To "exclude specific goals from `KafkaRebalance` resources"  we would just remove the goal from the `KafkaRebalance` resource itself. To make sure exclude a specific goal from being used in any goal list in Cruise Control we would remove it from the supported goals list

I think we should either remove this sentence or change it to something like:

```suggestion
To simplify configuration, use the inherited supported goals.
However, if you don't want specific goals to be supported or used for any proposal generation generation, remove them from the supported goals list. 
```
That being said, I think this  information is already covered by the first two sentences of this section, so I am for removing this line

PaulRMellor replied:
I think its simpler to remove. Removed.

kyguy commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -2,54 +2,508 @@
+Cruise Control uses default goals to generate an optimization proposal. 
+You can override default goals by setting proposal-specific optimization goals in a `KafkaRebalance` resource.
+
+If `default.goals` is not specified in the Cruise Control deployment configuration, Strimzi configures `default.goals` to the list of goals specified in `config.goals`. 
```
I believe the CC configuration is just `goals` although this field is in the `cruiseControl.config` section of the `Kafka` resource, just like `default.goals` is

```suggestion
If `default.goals` is not specified in the Cruise Control deployment configuration, Strimzi configures `default.goals` to the list of supported goals specified in `goals`. 
```

kyguy commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -2,54 +2,508 @@
+
+If `default.goals` is not specified in the Cruise Control deployment configuration, Strimzi configures `default.goals` to the list of goals specified in `config.goals`. 
+
+The optimization proposal based on this goals list is then generated and cached..
```
```suggestion
The optimization proposal based on this supported goals list is then generated and cached..
```

kyguy commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -2,54 +2,508 @@
+Goals labeled as hard are mandatory constraints that must be satisfied for optimization proposals.
+
+* `RackAwareGoal` (hard)
+* `MinTopicLeadersPerBrokerGoal`
```
```suggestion
* `MinTopicLeadersPerBrokerGoal` (hard)
```

kyguy commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -2,54 +2,508 @@
+With Cruise Control deployed for use in Strimzi, the process to generate and approve an optimization proposal is as follows:
+
+. Create a `KafkaRebalance` resource specifying optimization goals and any specific configurations. 
+This resource triggers Cruise Control to initiate the optimization process.
```
```suggestion
This resource triggers Cruise Control to initiate the optimization proposal generation process.
```

kyguy commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -2,54 +2,508 @@
+
+. Create a `KafkaRebalance` resource specifying optimization goals and any specific configurations. 
+This resource triggers Cruise Control to initiate the optimization process.
+. Strimzi Metrics Reporter runs in Kafka brokers, collecting raw metrics and publishing them to a dedicated Kafka topic (`strimzi.cruisecontrol.metrics`).
```
The Strimzi Metrics Reporter is a different metrics reporter
```suggestion
. Cruise Control Metrics Reporter runs in Kafka brokers, collecting raw metrics and publishing them to a dedicated Kafka topic (`strimzi.cruisecontrol.metrics`).
```

kyguy commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -2,54 +2,508 @@
+. Create a `KafkaRebalance` resource specifying optimization goals and any specific configurations. 
+This resource triggers Cruise Control to initiate the optimization process.
+. Strimzi Metrics Reporter runs in Kafka brokers, collecting raw metrics and publishing them to a dedicated Kafka topic (`strimzi.cruisecontrol.metrics`).
+Metrics for partition topics and modeling the impact of rebalances are collected in other xref:proc-cruise-control-auto-created-topics-{context}[topics automatically created when Cruise Control is deployed]. 
```
```suggestion
Metrics for brokers, topics, and partitions are aggregated, sampled and then stored in other xref:proc-cruise-control-auto-created-topics-{context}[topics automatically created when Cruise Control is deployed]. 
```

kyguy commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -2,54 +2,508 @@
+This resource triggers Cruise Control to initiate the optimization process.
+. Strimzi Metrics Reporter runs in Kafka brokers, collecting raw metrics and publishing them to a dedicated Kafka topic (`strimzi.cruisecontrol.metrics`).
+Metrics for partition topics and modeling the impact of rebalances are collected in other xref:proc-cruise-control-auto-created-topics-{context}[topics automatically created when Cruise Control is deployed]. 
+. Load Monitor collects the metrics from Kafka brokers, including CPU, disk, and network utilization data, and sends load data to the Analyzer and Anomaly Detector.
```
The load monitor is the component collecting, processing and storing the metrics mentioned in the previous sentence. It creates a cluster model of the load of the cluster which is used by the Analyzer and Anomaly Detector

kyguy commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -2,54 +2,508 @@
+Metrics for partition topics and modeling the impact of rebalances are collected in other xref:proc-cruise-control-auto-created-topics-{context}[topics automatically created when Cruise Control is deployed]. 
+. Load Monitor collects the metrics from Kafka brokers, including CPU, disk, and network utilization data, and sends load data to the Analyzer and Anomaly Detector.
+. Anomaly Detector continuously monitors the health and performance of the Kafka cluster, checking for things like broker failures or disk capacity issues, that could impact cluster stability.
+. Analyzer processes the collected metrics and anomalies to construct a _workload model_ of the current state of the Kafka cluster. 
```
The analyzer creates optimization proposals based on the workload models from the Load Monitor

kyguy commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -2,54 +2,508 @@
+. Load Monitor collects the metrics from Kafka brokers, including CPU, disk, and network utilization data, and sends load data to the Analyzer and Anomaly Detector.
+. Anomaly Detector continuously monitors the health and performance of the Kafka cluster, checking for things like broker failures or disk capacity issues, that could impact cluster stability.
+. Analyzer processes the collected metrics and anomalies to construct a _workload model_ of the current state of the Kafka cluster. 
+Based on configured goals and capacities, it generates an optimization proposal for balancing partition distribution across brokers.
```
I am not sure I understand the phrase "balancing partition distributions"? Is this a typo? 

Should it say "balancing partitions across brokers" or "evenly distributing partitions across brokers", or something along those lines?

kyguy commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -2,54 +2,508 @@
+However, any hard goals included in the proposal-specific goals will still be treated as hard goals by Cruise Control, even with `skipHardGoalCheck: true`.
+
+[id='con-optimization-proposals-{context}']
+== Optimization proposals
```
What's the motivation behind providing summary of how the internal components of Cruise Control work? Are there things we want to mention here that are not already mentioned in the upstream documentation? [1]

I agree with providing clarification/documentation surrounding to points that are Strimzi specific or unclear from upstream Cruise Control documentation. However, it seems some of the sentences here summarize Cruise Control components whose function is already documented/described in the upstream Cruise Control docs [1]. i.e the sentences surrounding the Load Monitor, Analyzer, and Anomaly Detector.

[1] https://github.com/linkedin/cruise-control/wiki/Overview

kyguy commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -2,54 +2,508 @@
+.Cruise Control optimization process
+image:kafka-concepts-cruise-control.png[Cruise Control process]
+
+Optimization proposals comprise separate partition reassignment commands.
```
```suggestion
Optimization proposals comprise of a list of partition reassignment mappings.
```

kyguy commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -2,54 +2,508 @@
+image:kafka-concepts-cruise-control.png[Cruise Control process]
+
+Optimization proposals comprise separate partition reassignment commands.
+When you approve a proposal, the Cruise Control server applies these commands to the Kafka cluster.
```
```suggestion
When you approve a proposal, the Cruise Control server applies these reassignments to the Kafka cluster.
```

PaulRMellor commented:
Thanks for latest comments @kyguy 
I've addressed them all as suggested.

Regarding the summary of Cruise Control components. Previously, we mentioned some but not all them without any context. I thought it would be useful at least to mention them all in summary and show how they operate within a Strimzi context, how requests are made and proposals generated etc. I think the diagram is useful (and good for SEO and user experience).

kyguy commented on documentation/modules/cruise-control/con-rebalance-performance.adoc:
```diff
@@ -3,67 +3,48 @@
 
-The relevant configurations are summarized in the following table.
+* Set Cruise Control server configurations in `Kafka.spec.cruiseControl.config` in the `Kafka` resource.
+* Set individual rebalances in `KafkaRebalance.spec` in the `KafkaRebalance` resource.
```
```suggestion
* Set proposal-specific configurations in `KafkaRebalance.spec` in the `KafkaRebalance` resource.
```

kyguy commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -2,54 +2,506 @@
+
+Use the summary to decide whether to approve or reject an optimization proposal.
+
+Approving an optimization proposal:: You approve the optimization proposal by setting the `strimzi.io/rebalance` annotation of the `KafkaRebalance` resource to `approve`.
```
Looks like there is a section `Approving optimization proposals`, should this be referenced and/or moved closer to auto approval section?

PaulRMellor commented:
Thanks for the comments @kyguy 
All pretty straightforward and addressed.
Doc is shaping up nicely

kyguy commented:
Hi @PaulRMellor, was to commit with the changes pushed here?

PaulRMellor commented:
> Hi @PaulRMellor, was the commit with the changes pushed here?

There now: https://github.com/strimzi/strimzi-kafka-operator/pull/10810/commits/d47f53a8af44901727bf7a2dd34a6700ae08888f

kyguy reviewed (approved):
Looks great @PaulRMellor, well done

PaulRMellor commented:
> Looks great @PaulRMellor, well done

Thanks for your help and applying your knowledge, @kyguy
